dot/network/request_response.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package network
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/gossamer/dot/peerset"
+	libp2pnetwork "github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/protocol"
+)
+
+type RequestResponseProtocol interface {
+	DoRequest(to peer.ID, req Message, res ResponseMessage) error
+}
+
+type requestResponseProtocol struct {
+	ctx             context.Context
+	host            *host
+	requestTimeout  time.Duration
+	maxResponseSize uint64
+	protocolID      protocol.ID
+	responseBufMu   sync.Mutex
+	responseBuf     []byte
+}
+
+func (rrp *requestResponseProtocol) DoRequest(to peer.ID, req Message, res ResponseMessage) error {
+	rrp.host.p2pHost.ConnManager().Protect(to, "")
+	defer rrp.host.p2pHost.ConnManager().Unprotect(to, "")
+
+	ctx, cancel := context.WithTimeout(rrp.ctx, rrp.requestTimeout)
+	defer cancel()
+
+	stream, err := rrp.host.p2pHost.NewStream(ctx, to, rrp.protocolID)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err := stream.Close()
+		if err != nil {
+			logger.Warnf("failed to close stream: %s", err)
+		}
+	}()
+
+	if err = rrp.host.writeToStream(stream, req); err != nil {
+		return err
+	}
+
+	return rrp.ReceiveResponse(stream, res)
+}
+
+func (rrp *requestResponseProtocol) ReceiveResponse(stream libp2pnetwork.Stream, msg ResponseMessage) error {
```

Comment:
receive response could be a private method since it will be used only by the `DoRequest` method
```suggestion
func (rrp *requestResponseProtocol) ReceiveResponse(stream libp2pnetwork.Stream, msg ResponseMessage) error {
```